include/scrimmage/common/ParameterServer.h

```diff
@@ -0,0 +1,138 @@
+            auto it_type = it_name->second.find(typeid(T).name());
+            if (it_type != it_name->second.end()) {
+                // This param name / type combo exists
+                // Has this specific plugin already registered this parameter?
+                for (const ParameterBasePtr &ptr : it_type->second) {
+                    if (ptr->owner() == owner) {
+                        return false;
+                    }
+                }
+            }
+            // Add the parameter to the params_ map
+            (it_name->second)[typeid(T).name()].push_back(param);
+        }
+        return true;
+    }
+
+    template <class T>
+    bool set_param(const std::string &name, const T &value) {
+        // Search for the parameter name in the server
+        auto it_name = params_.find(name);
+        if (it_name != params_.end()) {
+            // The param name exists, does the type exist?
+            auto it_type = it_name->second.find(typeid(T).name());
+            if (it_type != it_name->second.end()) {
+                // The param name/type exists in the list, iterate over the
+                // list of parameters and call their set_value functions
+                for (ParameterBasePtr param : it_type->second) {
+                    auto param_cast = std::dynamic_pointer_cast<Parameter<T>>(param);
+                    if (param_cast) {
+                        param_cast->set_value(value);
+                    }
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    template <class T>
+    bool unregister_param(const std::string &name, PluginPtr owner) {
+        // Search for the parameter name in the server
+        auto it_name = params_.find(name);
+        if (it_name != params_.end()) {
+            auto it_type = it_name->second.find(typeid(T).name());
+            if (it_type != it_name->second.end()) {
+                // The parameter name/type exists, let's find the specific
+                // plugin
+                auto it_param = std::remove_if(it_type->second.begin(),
+                                               it_type->second.end(),
+                                               [&](ParameterBasePtr ptr) {
+                                                   return ptr->owner() == owner; });
+                if (it_param != it_type->second.end()) {
+                    // The plugin created this param, erase it.
+                    it_type->second.erase(it_param, it_type->second.end());
+                    return true;
+                }
+            }
+        }
+        // Could not find the appropriate param name/type/plugin-owner
+        return false;
+    }
+
+ protected:
+    // Key 1: parameter name (string)
+    // Key 2: parameter type (as a string)
+    // Value: ParameterBase shared_ptr
+    std::map<std::string, std::map<std::string, std::list<ParameterBasePtr>>> params_;
```

Comment:
unordered_map may be more efficient here. Also, I think instead of ``list``, we should be using ``std::set``. This will get rid of the ``std::remove_if``` logic above.

Reply:
Updated.